### PR TITLE
[iOS] Fixed maxLength of TextInput based on glyph count

### LINKED ISSUE #10964: TextInput maxlength cuts Unicode caracters (Emojis)
Hello everyone,

yesterday i noticed that a TextInput has a problem displaying Emojis at the end of the fields maxlength. 
If i insert the "white happy man with black hair raising one hand" Emoji (iOS) at the position maxlength-1 it shows a square with an "A" in it instead.
If i insert the "white happy man with black hair raising one hand" Emoji (iOS) at the position maxlength-2 it shows a square with an "yellow happy woman with blonde hair raising one hand" in it instead.
If i insert the "white happy man with black hair raising one hand" Emoji (iOS) at the position maxlength-3 it shows a square with an "white happy woman with black hair raising one hand" in it instead.
If i insert the "white happy man with black hair raising one hand" Emoji (iOS) at the position maxlength-4 it shows a square with an "yellow happy woman with blonde hair raising one hand" and the square with an "A" in it instead.

Seems like the maxlength cuts the unicode sign at the end and causes to show different Emojis than inserted

* React Native version: 0.36
* Platform: iOS - tested on iOS 10.1.1

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary

Fixes #10964 #22463 .
We need changed `maxLength` to count glyph instead. Because like emoji, some language，it represents using multiple bytes, so if we limit using characters, would leads some weird bugs, like clear the line or glyph not show correctly.

cc. @cpojer .

## Changelog

[iOS] [Fixed] - Fixed maxLength of TextInput based on glyph count

## Test Plan

If we input one emoji, it needs to show correctly.
```
<TextInput maxLength={1} />
```

### CODE REVIEW

File: Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 1]
```diff
@@ -20,6 +20,25 @@
 #import "RCTTextAttributes.h"
 #import "RCTTextSelection.h"
 
+@interface NSString (RCTUtility)
```
Comment (shergin) [inline Libraries/Text/TextInput/RCTBaseTextInputView.m]: We have to decouple that into separate file(s).

**Comment by zhongwuzw** [comment]
@shegin Is this what we want? 🤔 I restructure the code and consolidate max length handler into separate function.

File: Libraries/Text/TextInput/RCTBaseTextInputView.m [hunk 2]
```diff
@@ -619,6 +610,38 @@ - (void)handleInputAccessoryDoneButton
 
 #pragma mark - Helpers
 
+- (void)trimInputTextInRange:(NSRange)range
```
Comment (shergin) [inline Libraries/Text/TextInput/RCTBaseTextInputView.m]: Not exactly.
This separate function is still a method (it mutates `self`) which does not solve the complexity problem.

Comment (zhongwuzw) [reply Libraries/Text/TextInput/RCTBaseTextInputView.m]: @shergin Yeah, actually, Swift already solved this glyphs issue, but OC, we seems need to do these heavy things by ourself. Have no idea how to move on. 😂

Comment (shergin) [reply Libraries/Text/TextInput/RCTBaseTextInputView.m]: Yeah, Swift's strings are character-aware but normal Objective-C are not. Yeah... but that's only String's properties, that is not related to UITextField and UITextView implementation, right? So, we already have two subclasses for those classes that enhance the behaviors and fix some issue. Maybe we can encapsulate that character-aware logic there?

Or, what if we move that `maxLength` feature implementation into those classes and try to make it totally transparent for all external logic?

**Comment by zhongwuzw** [comment]
@shergin 
Hi :) I may need your some inputs, is this the right direction you wanted? 🤔

**Comment by cpojer** [comment]
I thought about this some more and we have a ton of code that on the backend expects maxLength to adhere to the real length and not the human readable length. Changing the default now and only for iOS will likely cause a lot of issues for us at Facebook and others.

If we want to solve this I think we may want to add another prop that explicitly specifies the max length is based on the glyph count instead of the number of bytes and make people opt-in to use that prop in the cases where the human readable length matters to them.

**Comment by zhongwuzw** [comment]
@cpojer 👌 We already improved TextInput so much, let's wait to see wether any user wants it, then we can add it. :)

**Pull request closed without merging**